server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java

```diff
@@ -283,63 +296,86 @@ LeafBucketCollector getLeafCollector(Comparable<BytesRef> value, LeafReaderConte
-                if (currentValueIsSet == false) {
-                    if (dvs.advanceExact(doc)) {
-                        long ord;
-                        while ((ord = dvs.nextOrd()) != NO_MORE_ORDS) {
-                            if (term.equals(dvs.lookupOrd(ord))) {
-                                currentValueIsSet = true;
-                                currentValueOrd = ord;
-                                currentValueUnmapped = null;
-                                break;
-                            }
-                        }
-                    }
-                }
-                assert currentValueIsSet;
                 next.collect(doc, bucket);
             }
         };
     }
 
+    private static class SlotAndOrd implements Comparable<SlotAndOrd> {
+        final int slot;
+        final long ord;
+
+        SlotAndOrd(int slot, long ord) {
+            this.slot = slot;
+            this.ord = ord;
+        }
+
+        @Override
+        public int compareTo(SlotAndOrd other) {
+            long norm1 = ord < 0 ? -ord - 1 : ord;
+            long norm2 = other.ord < 0 ? -other.ord - 1 : other.ord;
+            int cmp = Long.compare(norm1, norm2);
+            return cmp == 0 ? Long.compare(ord, other.ord) : cmp;
+        }
+    }
+
     /**
      * Remaps ordinals when switching LeafReaders. It's possible that a term is not mapped for the new LeafReader,
      * in that case remember the term so that future remapping steps can accurately be done.
      */
     private void remapOrdinals(SortedSetDocValues oldMapping, SortedSetDocValues newMapping) throws IOException {
+        // speed up the lookups by sorting ordinals first
+        SlotAndOrd[] sorted = new SlotAndOrd[numSlots];
         for (int i = 0; i < numSlots; i++) {
-            final long oldOrd = valuesOrd.get(i);
-            if (oldOrd != Long.MIN_VALUE) {
-                final long newOrd;
-                if (oldOrd >= 0) {
-                    final BytesRef newVal = oldMapping.lookupOrd(oldOrd);
-                    newOrd = newMapping.lookupTerm(newVal);
-                    if (newOrd < 0) {
-                        setValueWithBreaking(i, BytesRef.deepCopyOf(newVal));
-                    }
-                } else {
-                    newOrd = newMapping.lookupTerm(valuesUnmapped.get(i));
-                    if (newOrd >= 0) {
-                        setValueWithBreaking(i, null);
-                    }
+            sorted[i] = new SlotAndOrd(i, valuesOrd.get(i));
+        }
+        Arrays.sort(sorted);
+
+        long lastOldOrd = Long.MIN_VALUE;
+        long lastNewOrd = Long.MIN_VALUE;
+        for (int i = 0; i < numSlots; i++) {
+            final long slot = sorted[i].slot;
+            final long oldOrd = sorted[i].ord;
+            if (oldOrd == Long.MIN_VALUE) {
+                continue;
+            }
+            final long newOrd;
+            if (lastOldOrd == oldOrd) {
+                newOrd = lastNewOrd;
```

Comment:
`lastOldOrd` is not set anywhere (so this optimization is effectively disabled :D). Also, should we only reuse the previous ord when `oldOrd >= 0`? In case where it's negative, we might have different values in valuesUnmapped.

Reply:
Good catch, I pushed https://github.com/elastic/elasticsearch/pull/78313/commits/18ee058c3e2862aa2f99f732e27b53a423ced38b and also updated the unmapped value when a duplicate is found